node/hyperbahn/rate_limiting.js

```diff
@@ -0,0 +1,187 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+
+var DEFAULT_SERVICE_QPS_LIMIT = 500;
+var DEFAULT_TOTAL_QPS_LIMIT = 1000;
+var DEFAULT_BUCKETS = 20;
+var TOTAL_QPS = '~~TotalQPS++';
+
+function RateLimiting(options) {
+    if (!(this instanceof RateLimiting)) {
+        return new RateLimiting(options);
+    }
+    var self = this;
+
+    assert(options.serviceDispatchHandler, 'serviceDispatchHandler is required for RateLimiting');
+    self.serviceDispatchHandler = options.serviceDispatchHandler;
+    self.egressNodes = options.serviceDispatchHandler.egressNodes;
+    self.timers = self.serviceDispatchHandler.channel.timers;
+
+    self.qpsLimits = options.qpsLimits || Object.create(null);
+    self.counters = Object.create(null);
+    self.buckets = options.buckets || DEFAULT_BUCKETS;
+    assert(self.buckets > 0 && self.buckets <= 1000, 'counter buckets should between (0 1000]');
+    self.refreshDelay = 1000 / self.buckets;
+    self.defaultServiceQpsLimit = options.defaultServiceQpsLimit || DEFAULT_SERVICE_QPS_LIMIT;
+    self.totalQpsLimit = options.totalQpsLimit || DEFAULT_TOTAL_QPS_LIMIT;
+    self.buckets = options.buckets || DEFAULT_BUCKETS;
+
+    self.refresh();
+}
+
+RateLimiting.prototype.type = 'tchannel.hyperbahn.rate-limiting';
+
+RateLimiting.prototype.refresh =
+function refresh() {
+    var self = this;
+
+    // update hte sliding window
+    var services = Object.keys(self.counters);
+    for (var i = 0; i < services.length; i++) {
+        var serviceName = services[i];
+        var counter = self.counters[serviceName];
+        var next = (counter.index + 1) % self.buckets;
+        if (counter.buckets[next]) {
+            // offset the bucket being moved out
+            counter.qps -= counter.buckets[next];
+        }
+        assert(counter.qps >= 0, 'qps shouls always be larger equal to 0');
+        counter.index = next;
+        counter.buckets[counter.index] = 0;
+        if (serviceName === TOTAL_QPS) {
+            // total QPS limit never change
+            continue;
+        }
+        var limit = self.qpsLimits[serviceName] || self.defaultServiceQpsLimit;
+        counter.qpsLimit = limit / self.egressNodes.kValueFor(serviceName);
+    }
+
+    self.refreshTimer = self.timers.setTimeout(
```

Comment:
not sure why though. I am not convinced why having a timer fires 20 times per seconds is slower than call timers.now() thousands of times per seconds. Actually, frequent calls to timers.now() is an anti-pattern in system programming since reading timers are expensive.